File: 3892-best-time-to-buy-and-sell-stock-v/best-time-to-buy-and-sell-stock-v.py
class Solution(object):
    def maximumProfit(self, prices, k):
        n=len(prices)
        if n<2 or k==0:
            return 0
        a=[0]*n
        for _ in range(1,k+1):
            b=[0]*n
            c=-2e18
            d=-2e18
            for i in range(1,n):
                e=a[i-2] if i>1 else 0
                c=max(c,e-prices[i-1])
                d=max(d,e+prices[i-1])
                f=max(prices[i]+c,-prices[i]+d)
                b[i]=max(b[i-1],f)
            a=b
        return a[-1]
        """
        :type prices: List[int]
        :type k: int
        :rtype: int
        """
        

File: 3892-best-time-to-buy-and-sell-stock-v/best-time-to-buy-and-sell-stock-v.cpp
class Solution {
public:
    long long maximumProfit(vector<int>& a, int k) {
        int n = a.size();
        long long f[n][k + 1][3];
        memset(f, 0, sizeof(f));
        for (int j = 1; j <= k; j++) {
            f[0][j][1] = -a[0];
            f[0][j][2] = a[0];
        }
        for (int i = 1; i < n; i++) {
            for (int j = 1; j <= k; j++) {
                f[i][j][0] = max({f[i - 1][j][0], f[i - 1][j][1] + a[i], f[i - 1][j][2] - a[i]});
                f[i][j][1] = max(f[i - 1][j][1], f[i - 1][j - 1][0] - a[i]);
                f[i][j][2] = max(f[i - 1][j][2], f[i - 1][j - 1][0] + a[i]);
            }
        }
        return f[n - 1][k][0];
    }
};
